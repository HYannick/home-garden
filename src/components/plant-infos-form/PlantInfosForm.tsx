/** @jsx jsx */
import React from 'react';
import { Field, Formik } from 'formik';
import { css, jsx } from '@emotion/core';
import { useTranslation } from 'react-i18next';
import styled from '@emotion/styled';
import plantInfosCreateValidation from '../../pages/plant-infos-create/plant-infos-create.validation';
import InputField from '../forms/InputField';
import { Button } from '../Button';
import { LevelData, PlantInfosProps } from '../../pages/plant-infos-create/PlantInfosCreate.types';
import TextAreaField from '../forms/TextAreaField';
import SelectField from '../forms/SelectField';
import { FormCreate } from './PlantInfosForm.styled';

interface PlantFormProps {
  onSubmit: any,
  initialValues: PlantInfosProps,
  submitLabel: string
}

export const Label = styled('span')`
    font-size: 2rem;
    color: ${({theme}) => theme.palette.warning.dark};
    margin-bottom: 1rem;
    display: block;
`;

export const Spacer = styled('div')`
  height: 2rem;
`;

const AreaField: React.FC<{ label: string, levelName: string, descriptionName: string }> = ({ label, levelName, descriptionName }) => {
  const { t } = useTranslation();

  const selectors = [
    {
      label: t('plant_infos_form.level_select.low'),
      value: LevelData.LOW,
    },
    {
      label: t('plant_infos_form.level_select.medium'),
      value: LevelData.MEDIUM,
    },
    {
      label: t('plant_infos_form.level_select.high'),
      value: LevelData.HIGH,
    },
  ];

  return (
    <FormCreate.Control>
      <Label>{label}</Label>
      <span>{t('plant_infos_form.field_level_label')}</span>
      <Field type="text" name={levelName} selectors={selectors} component={SelectField}/>
      <Spacer />
      <span>{t('plant_infos_form.field_tips_label')}</span>
      <Field name={descriptionName} component={TextAreaField}/>
    </FormCreate.Control>
  );
};


const PlantInfosForm: React.FC<PlantFormProps> = ({ onSubmit, initialValues, submitLabel }) => {
  const { t } = useTranslation();

  return (
    <Formik
      validate={(values) => plantInfosCreateValidation(values, t)}
      onSubmit={onSubmit}
      initialValues={initialValues}
      render={(props) => {
        const { errors, touched, isSubmitting } = props;
        return (
          <FormCreate>
            <FormCreate.Wrapper>
              <FormCreate.Title>
                <p>Plant infos</p>
              </FormCreate.Title>
              <div css={css`flex:1; padding: 0 2.5rem 0 2rem;`}>
                <FormCreate.HelperText>
                  {t('plant_infos_form.helper_text')}
                </FormCreate.HelperText>
                <FormCreate.Control>
                  <Field type="text" name="name" component={InputField} label={t('plant_infos_form.field_name')}/>
                  {errors.name && touched.name && <FormCreate.ErrorField>{errors.name}</FormCreate.ErrorField>}
                </FormCreate.Control>
                <FormCreate.Control>
                  <Field
                    type="text" name="latin_name" component={InputField}
                    label={t('plant_infos_form.field_latin_name')}/>
                </FormCreate.Control>
                <FormCreate.Control>
                  <Field type="text" name="family" component={InputField} label={t('plant_infos_form.field_family')}/>
                </FormCreate.Control>
                <AreaField
                  label={t('plant_infos_form.field_exposure')}
                  levelName="exposure_level"
                  descriptionName="exposure_description"/>
                <AreaField
                  label={t('plant_infos_form.field_temperature')}
                  levelName="temperature_level"
                  descriptionName="temperature_description"/>
                <AreaField
                  label={t('plant_infos_form.field_watering')}
                  levelName="watering_level"
                  descriptionName="watering_description"/>
                <FormCreate.Control>
                  <Label>{t('plant_infos_form.field_soil')}</Label>
                  <span>{t('plant_infos_form.field_soil_helper')}</span>
                  <Field name="temperature_description" component={TextAreaField}/>
                </FormCreate.Control>
                <FormCreate.Control>
                  <Label>{t('plant_infos_form.field_plant_tips')}</Label>
                  <span>{t('plant_infos_form.field_plant_tips_helper')}</span>
                  <Field name="tips" component={TextAreaField}/>
                </FormCreate.Control>
                <FormCreate.Control>
                  <FormCreate.ButtonWrapper>
                    <Button variant="primary" type="submit" disabled={isSubmitting}>
                      {submitLabel}
                    </Button>
                    <Button variant="danger" type="reset" disabled={isSubmitting}>
                      {t('plant_form.button.cancel')}
                    </Button>
                  </FormCreate.ButtonWrapper>
                </FormCreate.Control>
              </div>
            </FormCreate.Wrapper>
          </FormCreate>
        );
      }}
    />
  );
};

export default PlantInfosForm;
