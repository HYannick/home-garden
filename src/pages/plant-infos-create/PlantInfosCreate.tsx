/** @jsx jsx */
import React, { Fragment } from 'react';
import { jsx } from '@emotion/core';
import { useTranslation } from 'react-i18next';
import { withRouter } from 'react-router';
import ActionBar from '../../layout/ActionBar';
import { SideLayer } from '../../layout/SideLayer';
import PlantInfosForm from '../../components/plant-infos-form/PlantInfosForm';
import { LevelData, PlantInfosProps } from './PlantInfosCreate.types';

const PlantInfosCreate: React.FC = ({ history, location }: any) => {
  const { t } = useTranslation();
  const { state } = location;
  const initialValues: PlantInfosProps = {
    name: '',
    latin_name: '',
    family: '',
    description: '',
    exposure_level: LevelData.LOW,
    exposure_description: '',
    temperature_level: LevelData.LOW,
    temperature_description: '',
    watering_level: LevelData.LOW,
    watering_description: '',
    soil_type: '',
    tips: '',
    ...state && state.previousValues
  };
  const goToCreateForm = (values: PlantInfosProps) => history.push({pathname: '/create', state: {plantInfos: values}});

  return (
    <Fragment>
      <ActionBar title={t('plant_create_title')} path="/search"/>
      <PlantInfosForm onSubmit={goToCreateForm} initialValues={initialValues} submitLabel={t('plant_infos_form.button.submit')}/>
      <SideLayer fullHeight/>
    </Fragment>
  );
};

export default withRouter(PlantInfosCreate);
