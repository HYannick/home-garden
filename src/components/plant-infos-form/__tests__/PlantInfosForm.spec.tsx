import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup } from '@testing-library/react';
import { renderWithTheme } from '../../../theme-wrapper';
import PlantInfosForm from '../PlantInfosForm';
import { LevelData } from '../../../pages/plant-infos-create/PlantInfosCreate.types';

afterEach(cleanup);
describe('PlantInfosForm', () => {
  const defaultProps = {
    t: jest.fn(),
    initialValues: {
      name: '',
      latin_name: '',
      family: '',
      description: '',
      exposure_level: LevelData.LOW,
      exposure_description: '',
      temperature_level: LevelData.LOW,
      temperature_description: '',
      watering_level: LevelData.LOW,
      watering_description: '',
      soil_type: '',
      tips: ''
    },
    onSubmit: jest.fn(),
    submitLabel: 'Create',
  };

  it('should render properly', () => {
    const { container } = renderWithTheme(
      <PlantInfosForm {...defaultProps} />,
    );
    expect(container).toMatchSnapshot();
  });
});
