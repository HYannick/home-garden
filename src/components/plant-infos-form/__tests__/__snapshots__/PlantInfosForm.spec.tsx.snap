// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlantInfosForm should render properly 1`] = `
<div>
  <form
    class="css-udnwx8"
  >
    <div
      class="css-ov1ktg"
    >
      <div
        class="css-rkdae6"
      >
        <p>
          Plant infos
        </p>
      </div>
      <div
        class="css-5c6nel"
      >
        <p
          class="css-1aya0bk"
        >
          plant_infos_form.helper_text
        </p>
        <div
          class="css-1diq458"
        >
          <label
            class="css-npa6bc"
            for="name"
          >
            plant_infos_form.field_name
          </label>
          <input
            class="css-1qxlqm1"
            id="name"
            name="name"
            type="text"
            value=""
          />
        </div>
        <div
          class="css-1diq458"
        >
          <label
            class="css-npa6bc"
            for="latin_name"
          >
            plant_infos_form.field_latin_name
          </label>
          <input
            class="css-1qxlqm1"
            id="latin_name"
            name="latin_name"
            type="text"
            value=""
          />
        </div>
        <div
          class="css-1diq458"
        >
          <label
            class="css-npa6bc"
            for="family"
          >
            plant_infos_form.field_family
          </label>
          <input
            class="css-1qxlqm1"
            id="family"
            name="family"
            type="text"
            value=""
          />
        </div>
        <div
          class="css-1diq458"
        >
          <span
            class="css-1cblmog"
          >
            plant_infos_form.field_exposure
          </span>
          <span>
            plant_infos_form.field_level_label
          </span>
          <label
            class="css-npa6bc"
            for="exposure_level"
          />
          <div
            class="css-19or460"
          >
            <select
              id="exposure_level"
              name="exposure_level"
            >
              <option
                value="LOW"
              >
                plant_infos_form.level_select.low
              </option>
              <option
                value="MEDIUM"
              >
                plant_infos_form.level_select.medium
              </option>
              <option
                value="HIGH"
              >
                plant_infos_form.level_select.high
              </option>
            </select>
            <div
              class="css-gw542d"
            >
              <svg
                viewBox="0 0 10 4"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.625 3.5a1 1 0 0 1-1.25 0L2.226 1.781A1 1 0 0 1 2.851 0h4.3a1 1 0 0 1 .625 1.781z"
                />
              </svg>
            </div>
          </div>
          <div
            class="css-1kbost9"
          />
          <span>
            plant_infos_form.field_tips_label
          </span>
          <label
            class="css-npa6bc"
            for="exposure_description"
          />
          <textarea
            class="css-vygrie"
            id="exposure_description"
            name="exposure_description"
          />
        </div>
        <div
          class="css-1diq458"
        >
          <span
            class="css-1cblmog"
          >
            plant_infos_form.field_temperature
          </span>
          <span>
            plant_infos_form.field_level_label
          </span>
          <label
            class="css-npa6bc"
            for="temperature_level"
          />
          <div
            class="css-19or460"
          >
            <select
              id="temperature_level"
              name="temperature_level"
            >
              <option
                value="LOW"
              >
                plant_infos_form.level_select.low
              </option>
              <option
                value="MEDIUM"
              >
                plant_infos_form.level_select.medium
              </option>
              <option
                value="HIGH"
              >
                plant_infos_form.level_select.high
              </option>
            </select>
            <div
              class="css-gw542d"
            >
              <svg
                viewBox="0 0 10 4"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.625 3.5a1 1 0 0 1-1.25 0L2.226 1.781A1 1 0 0 1 2.851 0h4.3a1 1 0 0 1 .625 1.781z"
                />
              </svg>
            </div>
          </div>
          <div
            class="css-1kbost9"
          />
          <span>
            plant_infos_form.field_tips_label
          </span>
          <label
            class="css-npa6bc"
            for="temperature_description"
          />
          <textarea
            class="css-vygrie"
            id="temperature_description"
            name="temperature_description"
          />
        </div>
        <div
          class="css-1diq458"
        >
          <span
            class="css-1cblmog"
          >
            plant_infos_form.field_watering
          </span>
          <span>
            plant_infos_form.field_level_label
          </span>
          <label
            class="css-npa6bc"
            for="watering_level"
          />
          <div
            class="css-19or460"
          >
            <select
              id="watering_level"
              name="watering_level"
            >
              <option
                value="LOW"
              >
                plant_infos_form.level_select.low
              </option>
              <option
                value="MEDIUM"
              >
                plant_infos_form.level_select.medium
              </option>
              <option
                value="HIGH"
              >
                plant_infos_form.level_select.high
              </option>
            </select>
            <div
              class="css-gw542d"
            >
              <svg
                viewBox="0 0 10 4"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5.625 3.5a1 1 0 0 1-1.25 0L2.226 1.781A1 1 0 0 1 2.851 0h4.3a1 1 0 0 1 .625 1.781z"
                />
              </svg>
            </div>
          </div>
          <div
            class="css-1kbost9"
          />
          <span>
            plant_infos_form.field_tips_label
          </span>
          <label
            class="css-npa6bc"
            for="watering_description"
          />
          <textarea
            class="css-vygrie"
            id="watering_description"
            name="watering_description"
          />
        </div>
        <div
          class="css-1diq458"
        >
          <span
            class="css-1cblmog"
          >
            plant_infos_form.field_soil
          </span>
          <span>
            plant_infos_form.field_soil_helper
          </span>
          <label
            class="css-npa6bc"
            for="temperature_description"
          />
          <textarea
            class="css-vygrie"
            id="temperature_description"
            name="temperature_description"
          />
        </div>
        <div
          class="css-1diq458"
        >
          <span
            class="css-1cblmog"
          >
            plant_infos_form.field_plant_tips
          </span>
          <span>
            plant_infos_form.field_plant_tips_helper
          </span>
          <label
            class="css-npa6bc"
            for="tips"
          />
          <textarea
            class="css-vygrie"
            id="tips"
            name="tips"
          />
        </div>
        <div
          class="css-1diq458"
        >
          <div
            class="css-115f2m6"
          >
            <button
              class="css-1r4gcaj"
              type="submit"
            >
              Create
            </button>
            <button
              class="css-116s2bk"
              type="reset"
            >
              plant_form.button.cancel
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
